pkg/reconciler/sequence/sequence.go

```diff
@@ -331,3 +337,148 @@ func (r *Reconciler) removeUnwantedSubscriptions(ctx context.Context, seq *v1.Se
+
+	// Handle intermediate channel policies (skip the first channel as it's the input channel!)
+	for i := 1; i < len(channels); i++ {
+		expectedPolicy := resources.MakeEventPolicyForSequenceChannel(s, channels[i], subs[i-1])
+		existingPolicy, exists := existingPolicyMap[expectedPolicy.Name]
+
+		if exists {
+			if !equality.Semantic.DeepDerivative(expectedPolicy.Spec, existingPolicy.Spec) {
+				policiesToUpdate = append(policiesToUpdate, expectedPolicy)
+			}
+			delete(existingPolicyMap, expectedPolicy.Name)
+		} else {
+			policiesToCreate = append(policiesToCreate, expectedPolicy)
+		}
+	}
+
+	// Handle input channel policies
+	inputPolicies, err := r.prepareInputChannelEventPolicy(s, channels[0])
+	if err != nil {
+		return fmt.Errorf("failed to prepare input channel EventPolicies: %w", err)
+	}
+	for _, inputPolicy := range inputPolicies {
+		existingInputPolicy, exists := existingPolicyMap[inputPolicy.Name]
+		if exists {
+			if !equality.Semantic.DeepDerivative(inputPolicy.Spec, existingInputPolicy.Spec) {
+				policiesToUpdate = append(policiesToUpdate, inputPolicy)
+			}
+			delete(existingPolicyMap, inputPolicy.Name)
+		} else {
+			policiesToCreate = append(policiesToCreate, inputPolicy)
+		}
+	}
+
+	// Any remaining policies in the map should be deleted
+	for _, policy := range existingPolicyMap {
+		policiesToDelete = append(policiesToDelete, policy)
+	}
+
+	// Perform the actual CRUD operations
+	if err := r.createEventPolicies(ctx, policiesToCreate); err != nil {
+		return fmt.Errorf("failed to create EventPolicies: %w", err)
+	}
+	if err := r.updateEventPolicies(ctx, policiesToUpdate); err != nil {
+		return fmt.Errorf("failed to update EventPolicies: %w", err)
+	}
+	if err := r.deleteEventPolicies(ctx, policiesToDelete); err != nil {
+		return fmt.Errorf("failed to delete EventPolicies: %w", err)
+	}
+
+	return nil
+}
+
+// listEventPoliciesForSequence lists all EventPolicies (e.g. the policies for the input channel and the intermediate channels) created during reconcileKind that are associated with the given Sequence.
+func (r *Reconciler) listEventPoliciesForSequence(s *v1.Sequence) ([]*eventingv1alpha1.EventPolicy, error) {
+	labelSelector := labels.SelectorFromSet(map[string]string{
+		resources.SequenceChannelEventPolicyLabelPrefix + "sequence-name": s.Name,
+	})
+	return r.eventPolicyLister.EventPolicies(s.Namespace).List(labelSelector)
+}
+
+func (r *Reconciler) prepareInputChannelEventPolicy(s *v1.Sequence, inputChannel *eventingduckv1.Channelable) ([]*eventingv1alpha1.EventPolicy, error) {
+	matchingPolicies, err := auth.GetEventPoliciesForResource(
+		r.eventPolicyLister,
+		v1.SchemeGroupVersion.WithKind("Sequence"),
+		s.ObjectMeta,
+	)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get matching EventPolicies for Sequence: %w", err)
+	}
+
+	if len(matchingPolicies) == 0 {
+		return nil, nil
+	}
+
+	sort.Slice(matchingPolicies, func(i, j int) bool {
```

Comment:
Do sorting here as the order of the matching policies is indeterminant. it cause the unit test (when sequence has multiple eventpolicies) become flaky.

Reply:
```
   sequence_test.go:3420: Unexpected status update (-want, +got):
          &v1.Sequence{
          	TypeMeta:   {},
          	ObjectMeta: {Name: "test-sequence", Namespace: "test-namespace"},
          	Spec:       {Steps: {{Destination: {URI: &{Scheme: "http", Host: "example.com", Path: "0"}}}, {Destination: {URI: &{Scheme: "http", Host: "example.com", Path: "1"}}}}, ChannelTemplate: &{TypeMeta: {Kind: "InMemoryChannel", APIVersion: "messaging.knative.dev/v1"}, Spec: &{Raw: "{}"}}},
          	Status: v1.SequenceStatus{
          		Status: v1.Status{
          			ObservedGeneration: 0,
          			Conditions: v1.Conditions{
          				{Type: "Addressable", Status: "Unknown", Reason: "emptyAddress", Message: "addressable is nil", ...},
          				{Type: "ChannelsReady", Status: "Unknown", Reason: "ChannelsNotReady", Message: "Channels are not ready yet, or there are none", ...},
          				{
          					... // 1 ignored and 3 identical fields
          					Reason: "EventPoliciesNotReady",
          					Message: strings.Join({
          						"event policies test-sequence-",
        - 						"ep, test-sequence-additional-1-ep, test-sequence-additional-2",
        + 						"additional-1-ep, test-sequence-additional-2-ep, test-sequence",
          						"-ep are not ready",
          					}, ""),
          				},
          				{
          					... // 1 ignored and 3 identical fields
          					Reason: "EventPoliciesNotReady",
          					Message: strings.Join({
          						"event policies test-sequence-",
        - 						"ep, test-sequence-additional-1-ep, test-sequence-additional-2",
        + 						"additional-1-ep, test-sequence-additional-2-ep, test-sequence",
          						"-ep are not ready",
          					}, ""),
          				},
          				{Type: "SubscriptionsReady", Status: "Unknown", Reason: "SubscriptionsNotReady", Message: "Subscriptions are not ready yet, or there are none", ...},
          			},
          			Annotations: nil,
          		},
          		SubscriptionStatuses: {{Subscription: {Kind: "Subscription", Namespace: "test-namespace", Name: "test-sequence-kn-sequence-0", APIVersion: "messaging.knative.dev/v1", ...}, ReadyCondition: {Type: "Ready", Status: "Unknown", Reason: "NoReady", Message: "Subscription does not have Ready condition"}}, {Subscription: {Kind: "Subscription", Namespace: "test-namespace", Name: "test-sequence-kn-sequence-1", APIVersion: "messaging.knative.dev/v1", ...}, ReadyCondition: {Type: "Ready", Status: "Unknown", Reason: "NoReady", Message: "Subscription does not have Ready condition"}}},
          		ChannelStatuses:      {{Channel: {Kind: "InMemoryChannel", Namespace: "test-namespace", Name: "test-sequence-kn-sequence-0", APIVersion: "messaging.knative.dev/v1", ...}, ReadyCondition: {Type: "Ready", Status: "Unknown", Reason: "NoReady", Message: "Channel does not have Ready condition"}}, {Channel: {Kind: "InMemoryChannel", Namespace: "test-namespace", Name: "test-sequence-kn-sequence-1", APIVersion: "messaging.knative.dev/v1", ...}, ReadyCondition: {Type: "Ready", Status: "Unknown", Reason: "NoReady", Message: "Channel does not have Ready condition"}}},
          		... // 3 identical fields
          	},
          }
        
        Full: &{{ } {test-sequence  test-namespace    0 0001-01-01 00:00:00 +0000 UTC <nil> <nil> map[] map[] [] [] []} {[{{<nil> http://example.com/0 <nil> <nil>} <nil>} {{<nil> http://example.com/1 <nil> <nil>} <nil>}] &TypeMeta{Kind:InMemoryChannel,APIVersion:messaging.knative.dev/v1,} <nil>} {{0 [{Addressable Unknown  {2024-07-31 01:38:04.798862164 -0400 EDT m=+3.880229171} emptyAddress addressable is nil} {ChannelsReady Unknown  {2024-07-31 01:38:04.798862164 -0400 EDT m=+3.880229171} ChannelsNotReady Channels are not ready yet, or there are none} {EventPoliciesReady False  {2024-07-31 01:38:04.798862164 -0400 EDT m=+3.880229171} EventPoliciesNotReady event policies test-sequence-additional-1-ep, test-sequence-additional-2-ep, test-sequence-ep are not ready} {Ready False  {2024-07-31 01:38:04.798862164 -0400 EDT m=+3.880229171} EventPoliciesNotReady event policies test-sequence-additional-1-ep, test-sequence-additional-2-ep, test-sequence-ep are not ready} {SubscriptionsReady Unknown  {2024-07-31 01:38:04.798862164 -0400 EDT m=+3.880229171} SubscriptionsNotReady Subscriptions are not ready yet, or there are none}] map[]} [{{Subscription test-namespace test-sequence-kn-sequence-0  messaging.knative.dev/v1  } {Ready Unknown  {2024-07-31 01:38:04.798688281 -0400 EDT m=+3.880055291} NoReady Subscription does not have Ready condition}} {{Subscription test-namespace test-sequence-kn-sequence-1  messaging.knative.dev/v1  } {Ready Unknown  {2024-07-31 01:38:04.798688999 -0400 EDT m=+3.880056004} NoReady Subscription does not have Ready condition}}] [{{InMemoryChannel test-namespace test-sequence-kn-sequence-0  messaging.knative.dev/v1  } {Ready Unknown  {2024-07-31 01:38:04.798241015 -0400 EDT m=+3.879608024} NoReady Channel does not have Ready condition}} {{InMemoryChannel test-namespace test-sequence-kn-sequence-1  messaging.knative.dev/v1  } {Ready Unknown  {2024-07-31 01:38:04.798241621 -0400 EDT m=+3.879608619} NoReady Channel does not have Ready condition}}] {<nil>  <nil> <nil>} <nil> {[]}}}
    --- FAIL: TestAllCases/sequence_with_multiple_eventpolicies (0.14s)
```